File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = 0
    arrays = [[] for _ in range(n+1)]
    visited = [0] * (n+1)
    for i, j in edge:
        arrays[i].append(j)
        arrays[j].append(i)
    
    q = deque()
    q.append(1)
    visited[1] = 1
    
    while q:
        x = q.popleft()
        for y in arrays[x]:
            if not visited[y]:
                visited[y] = visited[x] + 1
                q.append(y)
    
    print(visited)
    
    for i in range(1, n+1):
        if visited[i]==max(visited):
            answer+=1
            
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    
    static class Pair {
        int y, level;
        
        public Pair(int y, int level) {
            this.y = y;
            this.level = level;
        }
        
        
    }
    static int[] visited;
    static ArrayList<ArrayList<Integer>> arr;
    static int max = 0;
    
    public int solution(int n, int[][] edge) {
        visited = new int[n+1];
        arr = new ArrayList<>();
        for (int i = 0; i<=n; i++) {
            arr.add(new ArrayList<>());
        }
        for (int[] now: edge) {
            int a = now[0];
            int b = now[1];
            arr.get(a).add(b);
            arr.get(b).add(a);
        }
        int answer = 0;
        bfs();
        for (int i = 1; i<=n; i++) {
            if (visited[i]==max) {
                answer++;
            }
        }
        return answer;
    }
    
    static void bfs() {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(1, 0));
        
        while (!q.isEmpty()) {
            Pair curr = q.poll();
            int y = curr.y;
            int level = curr.level;
            
            for (int i: arr.get(y)) {
                if (i!=1 && visited[i]==0) {
                    visited[i] = level+1;
                    q.add(new Pair(i, level+1));
                    max = Math.max(max, level+1);
                }
            }
            
        }
        
    }
}